Source/DafnyCore/Resolver/PreType/PreType.cs

```diff
@@ -0,0 +1,326 @@
+//-----------------------------------------------------------------------------
+//
+// Copyright by the contributors to the Dafny Project
+// SPDX-License-Identifier: MIT
+//
+//-----------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Linq;
+using Bpl = Microsoft.Boogie;
+
+namespace Microsoft.Dafny {
+  public abstract class PreType {
+    public PreType Normalize() {
+      var t = this;
+      while (t is PreTypeProxy proxy && proxy.PT != null) {
+        t = proxy.PT;
+      }
+      return t;
+    }
+
+    public static bool Same(PreType a, PreType b) {
+      a = a.Normalize();
+      b = b.Normalize();
+      if (a is DPreType ap && b is DPreType bp && ap.Decl == bp.Decl) {
+        Contract.Assert(ap.Arguments.Count == bp.Arguments.Count);
+        for (var i = 0; i < ap.Arguments.Count; i++) {
+          if (!Same(ap.Arguments[i], bp.Arguments[i])) {
+            return false;
+          }
+        }
+        return true;
+      }
+      return a == b;
+    }
+
+    public PreType UrAncestor(PreTypeResolver ptResolver) {
```

Comment:
Example of where comments would be helpful. I assume we following definitions to find typing constraints, and if so it would be good to say so and explain what assumptions we are making to avoid going in cycles

Reply:
I added a comment.

Regarding cycles, this is a good observation. The method as it stands assumes there are no cycles. I added a comment about this. Some of the cycle detection did not make it into this PR, but are part of new upcoming PRs.